app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt

```diff
@@ -1,33 +1,250 @@
+            CFGLinkType.UNCONDITIONAL,
+            IntermediateFormTreeNode.Call(functionLocationInCode)
+        )
+
+        var readResultNode: IFTNode? = null
+        if (variableToStoreFunctionResult != null) {
+            cfgBuilder.addLinkFromAllFinalRoots(
+                CFGLinkType.UNCONDITIONAL,
+                IntermediateFormTreeNode.RegisterWrite(Register.RAX, genRead(variableToStoreFunctionResult, true))
+            )
+            readResultNode = IntermediateFormTreeNode.RegisterRead(Register.RAX)
+        }
+        return FunctionDetailsGeneratorInterface.FunctionCallIntermediateForm(cfgBuilder.build(), readResultNode)
     }
 
     override fun genPrologue(): ControlFlowGraph {
-        return TODO()
+        val cfgBuilder = ControlFlowGraphBuilder()
+
+        // move args from Registers and Stack
+        for ((param, register) in parameters zip argPositionToRegister) {
+            cfgBuilder.addLinkFromAllFinalRoots(
+                CFGLinkType.UNCONDITIONAL,
+                genWrite(
+                    param,
+                    IntermediateFormTreeNode.RegisterRead(register),
+                    true
+                )
+            )
+        }
+        for (param in parameters.drop(argPositionToRegister.size)) {
+            cfgBuilder.addLinkFromAllFinalRoots(
+                CFGLinkType.UNCONDITIONAL,
+                genWrite(
+                    param,
+                    IntermediateFormTreeNode.StackPop(),
+                    true
+                )
+            )
+        }
+
+        // backup callee-saved registers
+        for (register in caleeSavedRegistersWithoutRSP.reversed())
+            cfgBuilder.addLinkFromAllFinalRoots(
+                CFGLinkType.UNCONDITIONAL,
+                IntermediateFormTreeNode.StackPush(IntermediateFormTreeNode.RegisterRead(register))
+            )
+
+        // update rbp
+        val movRbpRsp = IntermediateFormTreeNode.RegisterWrite(
+            Register.RBP,
+            IntermediateFormTreeNode.RegisterRead(Register.RSP)
+        )
+        cfgBuilder.addLinkFromAllFinalRoots(CFGLinkType.UNCONDITIONAL, movRbpRsp)
+
+        // allocate memory for local variables
+        val subRsp = IntermediateFormTreeNode.RegisterWrite(
+            Register.RSP,
+            IntermediateFormTreeNode.Subtract(
+                IntermediateFormTreeNode.RegisterRead(Register.RSP),
+                IntermediateFormTreeNode.Const(variablesTotalOffset.toLong())
+            )
+        )
+        cfgBuilder.addLinkFromAllFinalRoots(CFGLinkType.UNCONDITIONAL, subRsp)
+
+        // update display
+        val savePreviousRbp = IntermediateFormTreeNode.RegisterWrite(
+            previousDisplayEntryRegister,
+            IntermediateFormTreeNode.MemoryRead(
+                IntermediateFormTreeNode.Subtract(
+                    displayAddress,
+                    IntermediateFormTreeNode.Const((memoryUnitSize * depth).toLong())
```

Comment:
I think it's pretty compact right now. If we introduced a method like `genAddress(base, offset)` then we'd still have to pass it the same arguments as before so it will kind of look the same.

Reply:
No, what I meant was to extract something like `getDisplaySlotAddress` (without arguments, just for display).